tracker/derived.go

```diff
@@ -0,0 +1,235 @@
+package tracker
+
+import (
+	"github.com/vsariola/sointu"
+	"iter"
+	"slices"
+)
+
+/*
+	from modelData we can derive useful information that can be cached for performance
+	or easy access, because of nested iterations over the score or patch data.
+	i.e. this needs to update when the model changes, and only then.
+*/
+
+type (
+	derivedForUnit struct {
+		unit       *sointu.Unit
+		instrument *sointu.Instrument
+		sends      []*sointu.Unit
+	}
+
+	derivedForTrack struct {
+		instrumentRange          []int
+		tracksWithSameInstrument []int
+		title                    string
+	}
+
+	derivedModelData struct {
+		// map unit by ID
+		forUnit map[int]derivedForUnit
+		// map track by index
+		forTrack map[int]derivedForTrack
+	}
+)
+
+// public access functions
+
+func (m *Model) forUnitById(id int) *derivedForUnit {
+	forUnit, ok := m.derived.forUnit[id]
+	if !ok {
+		return nil
+	}
+	return &forUnit
+}
+
+func (m *Model) InstrumentForUnit(id int) *sointu.Instrument {
+	fu := m.forUnitById(id)
+	if fu == nil {
+		return nil
+	}
+	return fu.instrument
+}
+
+func (m *Model) UnitById(id int) *sointu.Unit {
+	fu := m.forUnitById(id)
+	if fu == nil {
+		return nil
+	}
+	return fu.unit
+}
+
+func (m *Model) SendTargetsForUnit(id int) []*sointu.Unit {
+	fu := m.forUnitById(id)
+	if fu == nil {
+		return nil
+	}
+	return fu.sends
+}
+
+func (m *Model) forTrackByIndex(index int) *derivedForTrack {
+	forTrack, ok := m.derived.forTrack[index]
+	if !ok {
+		return nil
+	}
+	return &forTrack
+}
+
+func (m *Model) TrackTitle(index int) string {
+	ft := m.forTrackByIndex(index)
+	if ft == nil {
+		return ""
+	}
+	return ft.title
+}
+
+// public getters with further model information
+
+func (m *Model) TracksWithSameInstrumentAsCurrent() []int {
+	currentTrack := m.d.Cursor.Track
+	return m.derived.forTrack[currentTrack].tracksWithSameInstrument
+}
+
+func (m *Model) CountNextTracksForCurrentInstrument() int {
+	currentTrack := m.d.Cursor.Track
+	count := 0
+	for t := range m.TracksWithSameInstrumentAsCurrent() {
+		if t > currentTrack {
+			count++
+		}
+	}
+	return count
+}
+
+// init / update methods
+
+func (m *Model) initDerivedData() {
+	m.derived = derivedModelData{
+		forUnit:  make(map[int]derivedForUnit),
+		forTrack: make(map[int]derivedForTrack),
+	}
+	m.updateDerivedScoreData()
+	m.updateDerivedPatchData()
+}
+
+func (m *Model) updateDerivedScoreData() {
+	for index, _ := range m.d.Song.Score.Tracks {
+		firstInstr, lastInstr, _ := m.instrumentRangeFor(index)
+		m.derived.forTrack[index] = derivedForTrack{
+			instrumentRange:          []int{firstInstr, lastInstr},
+			tracksWithSameInstrument: slices.Collect(m.tracksWithSameInstrument(index)),
+			title:                    m.buildTrackTitle(index),
+		}
+	}
+}
+
+func (m *Model) updateDerivedPatchData() {
+	for _, instr := range m.d.Song.Patch {
+		for _, unit := range instr.Units {
+			m.derived.forUnit[unit.ID] = derivedForUnit{
```

Comment:
While it's quite harmless that the maps are only generated once, it still can theoretically grow infinitely over time, if the unit IDs grow. Probably should just create a new map every update, or use [clear](https://tip.golang.org/ref/spec#Clear)

Reply:
clear() is done now